env/composite_env_wrapper.h

```diff
@@ -293,6 +293,10 @@ class CompositeEnvWrapper : public Env {
   // calls to env, and all file operations to fs
   explicit CompositeEnvWrapper(Env* env, FileSystem* fs)
       : env_target_(env), fs_env_target_(fs) {}
+  explicit CompositeEnvWrapper(Env* env, std::unique_ptr<FileSystem>&& fs)
+      : env_target_(env), fs_env_ptr_(std::move(fs)), fs_env_target_(fs_env_ptr_.get()) {}
+  explicit CompositeEnvWrapper(Env* env, std::shared_ptr<FileSystem> fs)
```

Comment:
EnvWrapper redirects to a base Env, CompositeEnvWrapper redirects to a base Env and a base FileSystem. EnvWrapper could be made to conditionally redirect file operations to an Env or a FileSystem depending on how its constructed, but I choose to keep the two separate.

Reply:
Why does an EnvWrapper need to do anything with a FileSystem?  I do not see the need for both classes and think it just adds an extra layer and more confusion.  Or is the plan to deprecate the EnvWrapper code?

Since an Env has a FileSystem, there is need to wrap the FileSystem or related calls in the EnvWrapper class (just like the CompositeEnvWrapper does).  You can create an EnvWrapper with just an Env (in which case it gets the FileSystem from the input arg) or with both an Env and a FileSystem.  In both cases, it passes the FileSystem to the Env parent.

Reply:
The plan is to deprecate the file related APIs in both Env and EnvWrapper eventually. 

You have a good point about wrapping the FileSystem related calls in EnvWrapper. Taking MockEnv as an example, it could be split into two -

```
class MockFileSystem : public FileSystemWrapper {
  MockFileSystem(std::shared_ptr<FileSystem> base_fs)
    :  FileSystemWrapper(base_fs) {}
```
and
```
class MockEnv : public EnvWrapper {
  MockEnv(Env* base_env, std::shared_ptr<FileSystem> base_fs)
    :  Env(std::make_shared<MockFileSystem>(base_fs)) {}
```
and FileSystem related calls in EnvWrapper would look something like this -
```
class EnvWrapper : public Env {
  EnvWrapper(Env* base_env, std::shared<FileSystem> base_fs)
    :  Env(base_fs), target_(base_env), constructed_with_fs_(true) {}

  Status NewSequentialFile(...) {
    if (constructed_with_fs_) {
      return file_system_->NewSequentialFile(...);
    } else {
      return target_->NewSequentialFile(...);
    }
  }
```

There may be other ways of doing it - need to think about it more. Will defer this to a separate PR. Right now, my focus is on porting an Env+FS implementation that inherits directly from Env, very similar to HdfsEnv.

Reply:
Why isn't the model:
- In the base env classes, the "legacy file system" calls are no longer pure virtual.  Instead, they are wrapped as "file_system->DoSomething()"
- There is a "LegacyEnvWrapper" that converts a legacy env into a file system.  
- The Env class has a constructor that takes a FileSystem as an argument
- The no-arg Env constructor creates a LegacyEnvWrapper from "this"
- The EnvWrapper ONLY wraps the non-legacy FileSystem calls.  All other calls are handled by the base Env class, which will redirect it to the underlying FileSystem
- The EnvWrapper was two constructors, one that takes an Env and one that takes an Env and a FileSystem.  

There is no need for three classes (CompositeWrapper, EnvWrapper, and Env) as they all overlap.  

And there should not need to be modifications to the base Posix class, other than to separate out the PosixFileSystem, but it sounds like that is already done.

I would also make the case that it is very hard to prove that what you done works until you have covered a few more of the cases -- like making a few of the existing implementations follow this model.  Ideally, I would think you would do ALL of them to prevent code rot...

Reply:
> * In the base env classes, the "legacy file system" calls are no longer pure virtual.  Instead, they are wrapped as "file_system->DoSomething()"
> * The EnvWrapper ONLY wraps the non-legacy FileSystem calls.  All other calls are handled by the base Env class, which will redirect it to the underlying FileSystem

I don't disagree. This may be what we end up with. However, I'm not comfortable with making the calls non-pure virtual yet until the new APIs are proven with a FileSystem implementation that actually uses some of the features of the new API (its WIP currently). MockEnv etc are not good candidates. I understand they need to be accommodated, but they're not useful in proving the new API.

CompositeEnvWrapper and PosixEnv are internal to RocksDB. If/when we go with the approach you suggested, it should be very cheap to refactor them.